### PR TITLE
feat: add service chapter exclusion rules for release notes generation

### LINKED ISSUE #298: Service chapter exclude
### Feature Description

Add support for label-based exclusion rules on individual service chapters. Each rule is a group of labels that must all be present on an issue (AND logic) for the issue to be excluded from that chapter. Multiple independent groups can be defined per chapter — groups are evaluated with OR logic (if any group matches, the issue is excluded).

### Problem / Opportunity

Service chapters such as "Closed Issues without Pull Request" currently collect all qualifying issues with no way to filter out noise. Teams use label combinations like `scope:security` + `type:tech-debt` to mark issues intentionally closed without a PR (e.g. false-positive security alerts). Without exclusion rules, these flood the chapter and obscure genuinely actionable items.

### Acceptance Criteria

1. Each service chapter definition accepts an optional `exclude` field containing a list of label groups.
2. A label group is a list of label strings — all labels in the group must be present on an issue for it to be excluded (AND logic within a group).
3. Multiple groups are supported per chapter; an issue matching any single group is excluded (OR logic across groups).
4. Issues excluded this way do not appear in any other chapter (silently dropped).
5. If `exclude` is omitted or empty, chapter behaviour is unchanged.
6. Invalid or unrecognised labels in `exclude` are ignored without error.
7. Configuration syntax and behaviour are documented with examples.

### Proposed Solution

Add a new action input `service-chapter-exclude` that maps service chapter titles to one or more label-exclusion groups. Each group is a list of labels that must all be present (AND); an issue matching any group is excluded from that chapter (OR across groups).

Example `action.yml` usage:

```yaml
- uses: AbsaOSS/generate-release-notes@v1
  with:
    service-chapter-exclude: |
      Closed Issues without Pull Request ⚠️:
        - [scope:security, type:tech-debt]
        - [scope:security, type:false-positive]

### Dependencies / Related

```markdown
- `docs/features/service_chapters.md` — service chapter configuration reference

### Additional Context

First concrete use case: the "Closed Issues without Pull Request" chapter in the AUL project receives large numbers of `[SEC][FP=...]` security false-positive issues (labelled `scope:security` + `type:tech-debt`) that are closed without a PR by design. These should be excludable without hiding other legitimate issues that genuinely lack a PR.

### PR DESCRIPTION
**Pull request by miroslavpojer**
## Overview

Adds label-based exclusion rules for service chapters, giving users fine-grained control over which issues/PRs appear in each service chapter based on their labels.

Previously the only way to remove entries from service chapters was to hide the entire chapter. With this change users can define per-chapter or global label rules using AND/OR logic:
- **Per-chapter**: a matching record is excluded from that chapter only; it may still appear in others.
- **Global (`"*"` key)**: a matching record is dropped from all service chapters entirely.
- Global rules take precedence over per-chapter rules.

**Changed files:**
- `action.yml` — new `service-chapter-exclude` input
- `release_notes_generator/action_inputs.py` — new `ActionInputs.get_service_chapter_exclude()` method with YAML parsing and validation
- `release_notes_generator/chapters/service_chapters.py` — new `exclude_rules` constructor parameter; new helpers `_is_globally_excluded`, `_is_excluded_from_chapter`, `_matches_any_group`
- `release_notes_generator/builder/builder.py` — wires the new input into `ServiceChapters`
- `release_notes_generator/utils/constants.py` — new constants `GLOBAL_EXCLUDE_KEY`, `SERVICE_CHAPTER_EXCLUDE`
- `docs/features/service_chapters.md` — new "Label-Based Exclusion Rules" section
- `docs/configuration_reference.md` — reference entry for the new input
- `tests/unit/release_notes_generator/chapters/test_service_chapters.py` — tests for per-chapter and global exclusion rules; two duplicate tests removed
- `tests/unit/release_notes_generator/test_action_inputs.py` — tests for `get_service_chapter_exclude()` parsing and validation

## Release Notes
- Added `service-chapter-exclude` action input: label-based exclusion rules (AND/OR logic) for filtering records out of individual or all service chapters.
- Added `"*"` global key support in `service-chapter-exclude` to drop matching records from every service chapter at once.

## Related
Closes #298 

## Summary by CodeRabbit

* **New Features**
  * Added `service-chapter-exclude` input to exclude records from service chapters using label-groups (AND within groups, OR across groups); supports a reserved "*" key for global exclusions and is honored when warnings are enabled.
  * Updated default issue row format to include progress ({progress}) in generated rows.

* **Documentation**
  * Added configuration reference and usage guidance explaining syntax, precedence, and behavior for per-chapter and global exclusions.

* **Tests**
  * Added unit tests covering parsing, exclusion semantics, precedence, and edge cases.

* **Chores**
  * CI workflow adjusted to exercise related output behavior.

### CODE REVIEW

File: docs/features/service_chapters.md [hunk 1]
```diff
@@ -116,6 +117,46 @@ Or use comma-separated format:
 
 **Note:** Title matching is exact and case-sensitive. The `service-chapter-order` input is independent of `hidden-service-chapters`; hidden chapters are still hidden even if listed in the order.
 
+### Label-Based Exclusion Rules
+Use `service-chapter-exclude` to filter out issues from service chapters by label combinations. Each rule is a group of labels that must **all** be present on an issue (AND logic) for the issue to be excluded. Multiple groups per chapter are evaluated with OR logic (any group match excludes the issue).
+
+**Per-chapter exclusion** — excludes a matching record from that chapter only:
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply docs/features/service_chapters.md:123]: Fixed.

File: release_notes_generator/builder/builder.py [hunk 2]
```diff
@@ -72,6 +72,7 @@ def build(self) -> str:
                 used_record_numbers=self.custom_chapters.populated_record_numbers_list,
                 hidden_chapters=ActionInputs.get_hidden_service_chapters(),
                 chapter_order=ActionInputs.get_service_chapter_order(),
+                exclude_rules=ActionInputs.get_service_chapter_exclude(),
```
Comment (tmikula-dev) [inline release_notes_generator/builder/builder.py]: Just stood up, that it is not named chapter_exclude as are other with same  patterns. So or they are patterns and the action input should be named as service_exclude_rules or it is correct and the var should be chapter_exclude.

Comment (miroslavpojer) [reply release_notes_generator/builder/builder.py]: I messed this difference. Good catch, fixed.

File: tests/unit/release_notes_generator/chapters/test_service_chapters.py [hunk 3]
```diff
@@ -227,3 +227,96 @@ def test_chapter_order_none_uses_default():
...
+
+def test_populate_global_partial_and_failure(record_with_issue_closed_no_pull):
+    """'*' AND failure -> record not excluded."""
+    sc = ServiceChapters(
+        user_defined_labels=["bug", "enhancement"],
+        chapter_exclude={"*": [["label1", "nonexistent"]]},
+    )
+    sc.populate({1: record_with_issue_closed_no_pull})
+    assert 1 == len(sc.chapters[CLOSED_ISSUES_WITHOUT_PULL_REQUESTS].rows)
+
```
Comment (tmikula-dev) [inline tests/unit/release_notes_generator/chapters/test_service_chapters.py:281]: Just a nitpick, but these tests are just different scenarios of the same feature. Could be refactored into param unit test.

**Review by tmikula-dev — COMMENTED** [review]
Please see the comments.

File: release_notes_generator/chapters/service_chapters.py [hunk 4]
```diff
@@ -107,6 +109,13 @@ def __init__(
 
         self.show_chapter_merged_prs_linked_to_open_issues = True
 
+        # Exclude rules: split "*" (global) from per-chapter entries
+        all_rules = chapter_exclude if chapter_exclude is not None else {}
```
Comment (tmikula-dev) [inline release_notes_generator/chapters/service_chapters.py]: now you renamed into chapter_exclude, the semantic of now having all_rules does not make much sense.

File: tests/unit/release_notes_generator/chapters/test_service_chapters.py [hunk 5]
```diff
@@ -227,3 +227,96 @@ def test_chapter_order_none_uses_default():
...
+
+def test_populate_per_chapter_or_logic_second_group_matches(record_with_issue_closed_no_pull):
+    """OR logic: second group match is sufficient for exclusion."""
+    sc = ServiceChapters(
+        user_defined_labels=["bug", "enhancement"],
+        chapter_exclude={CLOSED_ISSUES_WITHOUT_PULL_REQUESTS: [["nonexistent"], ["label1", "label2"]]},
+    )
+    sc.populate({1: record_with_issue_closed_no_pull})
+    assert 0 == len(sc.chapters[CLOSED_ISSUES_WITHOUT_PULL_REQUESTS].rows)
```
Comment (tmikula-dev) [inline tests/unit/release_notes_generator/chapters/test_service_chapters.py]: These are three tests could be combined into one param. Behind there are several, that does also same check. The first one if is default behaviour is extra. But this is just if you want to refactor the tests a little bit the logic is correct. It is up to you.

**Review by tmikula-dev — COMMENTED** [review]
Please see the comments.

File: release_notes_generator/chapters/service_chapters.py (hunk 4 above)
Comment (miroslavpojer) [reply release_notes_generator/chapters/service_chapters.py]: Improved.

File: tests/unit/release_notes_generator/chapters/test_service_chapters.py (hunk 5 above)
Comment (miroslavpojer) [reply tests/unit/release_notes_generator/chapters/test_service_chapters.py]: Fixed.

**Review by tmikula-dev — APPROVED** [review]
Approving.
